[azurerm_database_vulnerability_assessment_rule_baseline] New resource for vulnerability scan baseline settings

benjamin37 opened the pull request:
related to #4161

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+				ValidateFunc: validation.StringInSlice([]string{
+					string(sql.VulnerabilityAssessmentPolicyBaselineNameDefault),
+					string(sql.VulnerabilityAssessmentPolicyBaselineNameMaster),
+				}, true),
```
Could we make this case sensitive:
```suggestion
				}, false),
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
```
Should we validate that this is not am empty string?

benjamin37 replied:
Good idea, I changed that

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	parameters := expandBaselineResults(d.Get("baseline_result").(*schema.Set))
+
+	result, err := client.CreateOrUpdate(ctx, resourceGroupName, serverName, databaseName, ruleId, *baselineName, *parameters)
+
```
For consistency with the rest of the provider code:

```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	id, err := parseAzureResourceID(*result.ID)
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+		return err
+	}
+
+	// Currently REST API always returns "Default" as baseline that's why the Id is created manually here as a workaround
```
Is there an open issue on the SDK to track this? if not might be worth opening one

benjamin37 replied:
This is not an issue of the SDK, it's an issue of the Azure REST api. I contacted the Azure support about this. They confirmed it's a bug and will be fixed in the future. Should we still open an issue on the SDK to track this?

katbyte replied:
🤔 

We usually just throw issues there so they can route it to the correct team (rather then us trying to track down the right contact)

But this does post a little bit of an issue if we are going to support multiple database types. I think you'll need to change the path depending on if its mysql/mssql/postgres ect?

benjamin37 replied:
I created an Issue https://github.com/Azure/azure-sdk-for-go/issues/5435

Sorry I was wrong, MS SQL is the only supported database type.

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	log.Printf("[INFO] Reading Azure ARM Vulnerability Assessment Rule Baselines.")
+
+	id, err := parseAzureResourceID(d.Id())
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	baselineName := expandBaselineName(id.Path["baselines"])
+
+	result, err := client.Get(ctx, resourceGroupName, serverName, databaseName, ruleId, *baselineName)
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	log.Printf("[INFO] Deleting Azure ARM Vulnerability Assessment Rule Baselines.")
+
+	id, err := parseAzureResourceID(d.Id())
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+	baselineName := expandBaselineName(id.Path["baselines"])
+
+	result, err := client.Delete(ctx, resourceGroupName, serverName, databaseName, ruleId, *baselineName)
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func expandBaselineName(baselineName string) *sql.VulnerabilityAssessmentPolicyBaselineName {
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,277 @@
+
+func testCheckAzureRMStorageAccountDatabaseDestroy(s *terraform.State) error {
+	err := testCheckAzureRMStorageAccountDestroy(s)
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,277 @@
+%s
+
+resource "azurerm_database_vulnerability_assessment_rule_baseline" "test" {
+  resource_group_name = "${azurerm_resource_group.test.name}"
```
Could we format this TF and align the assignments?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,277 @@
+%s
+
+resource "azurerm_database_vulnerability_assessment_rule_baseline" "test" {
+  resource_group_name = "${azurerm_resource_group.test.name}"
```
And this tf

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,277 @@
+%s
+
+resource "azurerm_database_vulnerability_assessment_rule_baseline" "test" {
+  resource_group_name = "${azurerm_resource_group.test.name}"
```
And this tf

benjamin37 replied:
done

katbyte commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
````diff
@@ -0,0 +1,159 @@
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "test" {
````
Could we use example instead of test for resource names?
```suggestion
resource "azurerm_resource_group" "example" {
```

benjamin37 replied:
done

katbyte commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,159 @@
+
+## Argument Reference
+
+The following arguments are supported:
```
Could we update the properties to reflect when changing them would create a new resource?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+			output := map[string]interface{}{}
+
+			if result := baselineResult.Result; result != nil {
+
```
```suggestion
```

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,238 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateMsSqlServerName,
```
Is this only for MS SQL servers & databases?

benjamin37 replied:
Ups, you are right this is not only a feature of MS SQL databases & servers. I changed the validation to "No Empty Strings".

benjamin37 commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
````diff
@@ -0,0 +1,159 @@
+## Example Usage
+
+```hcl
+resource "azurerm_resource_group" "example" {
````
Sorry, I guess I was wrong. Vulnerability Scans are only supported for "Azure SQL Database, Azure SQL Managed Instance and Azure SQL Data Warehouse". So I guess that MS SQL is the only supported one, e. g. postgres does not seem to have this feature.

https://docs.microsoft.com/en-us/azure/sql-database/sql-vulnerability-assessment

katbyte replied:
nah, we can wait untill that is supported in terraform

benjamin37 commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,306 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMDatabaseVulnerabilityAssessmentRuleBaseline_basic(t *testing.T) {
```
Sorry for the confusion this feature is only available for MS SQL databases.

benjamin37 commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,159 @@
+
+* `resource_group_name` - (Required) The name of the resource group that contains the sql server. Changing this forces a new resource to be created.
+
+* `server_name` - (Required) Specifies the name of the SQL Server. Changing this forces a new resource to be created.
```
Sorry I was wrong, MS SQL is the only supported database type. Should I still change this to database?

katbyte replied:
ahh nah, instead we should change it to `MS SQL Server` :

benjamin37 replied:
Done

benjamin37 commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,159 @@
+
+# azurerm_database_vulnerability_assessment_rule_baseline
+
+Manages a Database Vulnerability Assessment Rule Baseline.
```
I have added a note that it's only compatible with MS SQL databases.

benjamin37 commented on website/docs/r/database_vulnerability_assessment_rule_baseline.html.markdown:
````diff
@@ -0,0 +1,159 @@
+Database Vulnerability Assessment Rule Baseline can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_database_vulnerability_assessment_rule_baseline.example  /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/acceptanceTestResourceGroup1/providers/Microsoft.Sql/servers/mysqlserver47547/databases/mysqldatabase/vulnerabilityAssessments/Default/rules/VA2065/baselines/master 
````
For managed instances the id will look like this:

/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/managedInstances/{managedInstanceName}/databases/{databaseName}/vulnerabilityAssessments/default/rules/{ruleId}/baselines/{baselineName}

I leave this conversation open until we decide to include managed databases or not.

katbyte replied:
Sounds good

benjamin37 commented:
Hi @katbyte 

sorry I was wrong with the statement that multiple database types are supported for vulnerability assessment feature. The only other supported database type except "MS SQL" is "Azure Managed Database" (see [sql-vulnerability-assessment](https://docs.microsoft.com/en-us/azure/sql-database/sql-vulnerability-assessment)). Since managed instances are not supported by the azurerm provider I would also rather not support this resource for managed databases. What do you think?

Kind regards
Benjamin

katbyte commented on azurerm/resource_arm_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,232 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
We can change this back to validate mssql database name

benjamin37 replied:
Reverted it back

katbyte commented:
@benjamin37,

Thanks for looking into that. For now just supporting mssql is fine. However since that is the case could we rename this resource to `azurerm_smsql_database_vulnerability_assessment_rule_baseline` to accurtly depict that?

Once that's done i think this should be go !

benjamin37 commented:
Hi @katbyte,

thanks for your reply! I renamed the resources and made the names and descriptions MS SQL specific.

katbyte commented on website/docs/r/mssql_database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,161 @@
+}
+
+resource "azurerm_mssql_database_vulnerability_assessment_rule_baseline" "example" {
+  resource_group_name = "${azurerm_resource_group.example.name}"
```
Could we align these assignments?

benjamin37 replied:
^done

katbyte commented on website/docs/r/mssql_database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,161 @@
+}
+
+resource "azurerm_template_deployment" "vulnerability_scan_settings" {
+  name = "accteststoragevss"
```
And these

benjamin37 replied:
done

katbyte commented on azurerm/internal/services/sql/client.go:
```diff
@@ -2,17 +2,19 @@ package sql
 
 import (
 	"github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/2015-05-01-preview/sql"
+	vulnerabilitySvc "github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/2017-03-01-preview/sql"
```
I believe we have and updated version of the API already included under mssql service package, will that one work or do we need to use a 3rd?

benjamin37 replied:
The `DatabaseVulnerabilityAssessmentRuleBaselinesClient` is only available in the 2017-03-01 version of the api. In the updated one there is just the managed version of this client.

See the REST API reference of Azure:
[Database Vulnerability Assessment Rule Baselines](https://docs.microsoft.com/en-us/rest/api/sql/databasevulnerabilityassessmentrulebaselines)
[Managed Database Vulnerability Assessment Rule Baselines](https://docs.microsoft.com/en-us/rest/api/sql/manageddatabasevulnerabilityassessmentrulebaselines)

Is it okay for you to leave the client in this file or should we move it to the mssql service package?

katbyte replied:
well that is.. unfortunate -_- I presume the manage version is for managed instances?

If you don't mind moving it mssql is the place it should live as we are going to slowly retire the sql package over time as we udpate the resources to use the newer API.

benjamin37 replied:
Yes, it's the version for managed instances. I moved it to the mssql clients.

katbyte commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,232 @@
+	return nil
+}
+
+func expandBaselineName(baselineName string) *sql.VulnerabilityAssessmentPolicyBaselineName {
```
I don't believe this function is necessary, could we just inline it?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,306 @@
+					resource.TestCheckResourceAttr(resourceName, "baseline_result.#", "1"),
+				),
+			},
+			{
```
Could we add an import check here?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,306 @@
+					resource.TestCheckResourceAttr(resourceName, "baseline_name", "master"),
+					resource.TestCheckResourceAttr(resourceName, "baseline_result.#", "2"),
+				),
+			},
```
and another import check here too

benjamin37 replied:
done

benjamin37 commented:
Hi @katbyte,

thanks for the reply! I have refactored the code like you suggested. 

I could reproduce the error in the test that you mentioned. Most of the time the test passes, it seems to be a timing issue. Have you tried executing the test multiple times?

The resource `azurerm_template_deployment.vulnerability_scan_settings` activates Advanced Data Security on the SQL server and specifies the storage container where the vulnerability scan results should be saved to. This is basically the pre-requirement to perform vulnerability scans and set baseline settings for the vulnerability scans. The error sounds like the Advanced Data Security settings are not active when we try to set the baseline settings even if there is an explicit `depends_on`. 

Do you have any idea how to make sure that the Advanced Data Security settings are already live before creating the baseline settings? Is this a general problem with ARM Templates or the mix of plain terraform and ARM Templates?

Kind regards
Benjamin

benjamin37 commented:
Hi @katbyte,

I guess I fixed the timing issue. It was between the two resources in `azurerm_template_deployment.vulnerability_scan_settings` and I added a `dependsOn` in the ARM template. Please have a check if the test works for you now. There shouldn't be any special configuration necessary for it to run.

benjamin37 commented:
Hi @katbyte,

there is no particular reason for not writing the resources I used as ARM Templates as terraform resources. For now I just didn't have the time for it. Maybe I can do this in a future pull request and remove the ARM templates in the example and test here. Would this be fine for you?

benjamin37 commented:
Hi @katbyte,

I thought I responded to all you remarks with my last comment. Did I miss anything?

Kind regards
Benjamin

katbyte commented:
Hi @benjamin37,

The one thing left for this PR is to remove the testy dependency on the ARM template and write that resource. I was hoping to do that myself but other things got in the way. Upon internal conversation we've re-affirmed our desire to have all tests only use terraform resources and not arm templates.

And yes, travis is failing do to unrelated issues with the apache git.

benjamin37 commented:
Hi @katbyte,

thanks for the feedback! I'm working on the other resources which are currently used as arm templates in the examples.

Kind regards
Benjamin

benjamin37 commented:
Hi @katbyte, 

I created a new resource for the security alert policy "azurerm_mssql_server_security_alert_policy". The other resource I used in the arm template is server vulnerability assessments. This resources uses api version 2018-06-01-preview (https://docs.microsoft.com/en-us/rest/api/sql/servervulnerabilityassessments) which is currently not available in the azure sdk for go, see https://github.com/Azure/azure-sdk-for-go/tree/master/services/preview/sql/mgmt). Is it okay for you to live with this 1 arm template resource until the azure sdk for go supports the 2018-06-01-preview api version?

Kind regards
Benjamin

mybayern1974 commented:
@benjamin37, an updated Go sdk v33.2.0 just got released [https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md
](https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md), which contains the 2018-06-01-preview sdk: [https://github.com/Azure/azure-sdk-for-go/tree/master/services/preview/sql/mgmt/2018-06-01-preview/sql](https://github.com/Azure/azure-sdk-for-go/tree/master/services/preview/sql/mgmt/2018-06-01-preview/sql), you may wanna have a try to see whether it could be used to replace the ARM template work around. Of cource once things could be proved to work, there will be an extra effort to integrate this SDK to this repo.

benjamin37 commented:
Hi @mybayern1974,

thanks for push this update in the go sdk. I guess this should work.

@katbyte should the azure sdk for go be updated in this pull request or should this be done in a separate one?

katbyte commented:
@benjamin37, we have updated master to v33.2.0 so i think your going to have to rebase or merge

benjamin37 commented:
@katbyte I did the merge.

benjamin37 commented:
Hi @katbyte & @tombuildsstuff,

please let me know when you want to have a look on this pr. I'll resolve the conflics in this case.

Kind regards
Benjamin

katbyte commented on azurerm/config.go:
```diff
@@ -247,7 +247,7 @@ func getArmClient(authConfig *authentication.Config, skipProviderRegistration bo
 	client.mssql = mssql.BuildClient(o)
 	client.msi = msi.BuildClient(o)
-	client.mysql = mysql.BuildClient(o)
+	client.mssql = mssql.BuildClient(o)
```
this is the same line as 249?

benjamin37 replied:
Yes, meanwhile someone else fixed this on master already. I'll remove this again.

benjamin37 commented:
Hi @katbyte,

I'm not sure where to find all API versions available for the Azure REST APIs. Can you tell me where to find it? These versions are displayed in their documentation:

API Version: 2017-03-01-preview:
 
* [Database Vulnerability Assessment Rule Baselines](https://docs.microsoft.com/en-us/rest/api/sql/databasevulnerabilityassessmentrulebaselines)
* [Server Security Alert Policies](https://docs.microsoft.com/en-us/rest/api/sql/serversecurityalertpolicies)

API Version: 2017-10-01-preview: 

* [Elastic Pools](https://docs.microsoft.com/en-us/rest/api/sql/elasticpools)

API Version: 2018-06-01-preview: 

* [Server Vulnerability Assessments](https://docs.microsoft.com/en-us/rest/api/sql/servervulnerabilityassessments)

---

In the **Azure SDK for go** the following versions are available:

API Version: 2017-03-01-preview:

* [Elastic Pools](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2017-03-01-preview/sql/elasticpools.go)
* [Database Vulnerability Assessment Rule Baselines](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2017-03-01-preview/sql/databasevulnerabilityassessmentrulebaselines.go)
* [Server Security Alert Policies](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2017-03-01-preview/sql/serversecurityalertpolicies.go)

API Version: 2017-10-01-preview: 

* [Elastic Pools](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2017-10-01-preview/sql/elasticpools.go)


API Version: 2018-06-01-preview: 

* [Elastic Pools](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2018-06-01-preview/sql/elasticpools.go)

* [Server Vulnerability Assessments](https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sql/mgmt/2018-06-01-preview/sql/servervulnerabilityassessments.go)

---

So we could for example increment the version for the elastic pool to api version 2018-06-01-preview to get rid of the 2017-10-01-preview version. What do you think?

Kind regards
Benjamin

katbyte commented:
*cool*

@benjamin37, i think that is the best path forward.  `2018-06-01-preview` for what we can and  `2017-03-01-preview` for everything else. hopefully at a future date there will be a version that supports everything!

benjamin37 commented:
Hi @katbyte,

I just saw that the 2018-06-01-preview ElasticPoolClient does not contain all the methods that the 2017 version has, e. g. essential ones like `Get`. I don't see a way to reduce the number of sql versions in the mssql Client.

katbyte commented:
@benjamin37, 

I have opened an [issue](https://github.com/Azure/azure-sdk-for-go/issues/6276) on the SDK as well as spoke to the service team. it's on the roadmap with no timeline yet so let's just get this merged in the best we can. Read to review and merge once the merge conflicts are resolved 🙂

benjamin37 commented:
Hi @katbyte,

thanks for creating the issue and catching up on this topic again.

I merged master branch and resolved the conflicts, you can have an other look on it.

benjamin37 commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,223 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
You mean `ValidateFunc: azure.ValidateResourceID,`? I'm not sure what this is doing exactly.

katbyte replied:
it basically makes sure the value is an azure resource ID

benjamin37 replied:
Are you sure the `rule_id` value should be an azure resource ID? Like it's currently build the `rule_id` has just values like `"VA2002"`. This is no azure resource id right? Azure resource id's should have an other format, e. g. for a key vault: `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/my-rg-414/providers/Microsoft.KeyVault/vaults/my-key-vault-323`.
When I add the validation `ValidateFunc: azure.ValidateResourceID` here, I'll get the error:

> Can not parse "rule_id" as a resource id: Cannot parse Azure ID: parse VA2002: invalid URI for request

katbyte replied:
Thanks for checking

benjamin37 commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,276 @@
+
+  depends_on = [
+    azurerm_mssql_server_vulnerability_assessment.test
+  ]
```
The api doesn't give any information about this. If you have a look at the UI in the Azure Portal there is `Advanced data security` on the SQL server that needs to be enabled to do any `Vulnerability Assessment Settings` (greyed out). 

In the `Advanced Data Security` settings of the SQL database you get the info

> Turn on Advanced Data Security for all databases on this server, at the cost of 12.02 EUR/server/month. This includes Data Discovery & Classification, Vulnerability Assessment and Advanced Threat Protection for the server. We invite you to a trial period for the first 30 days, without charge.

if `Advanced Data Security` is not enabled on the server. If it is enabled and you didn't specify the Storage Account where the scan results will be saved to in `Vulnerability Assessment Settings`, you will get the follow warning.

> Complete Advanced Data Security setup by selecting a storage account for Vulnerability Assessment

If you do API calls without respecting these dependencies you will get errors accordingly. So 

* `Advanced data security` on SQL *server* is needed for `Vulnerability Assessment Settings` on SQL *server*
* `Vulnerability Assessment Settings` on SQL *server* is needed to perform vulnerability scans on the *database* and do basline settings on the database
* `Vulnerability Assessment Settings` on SQL *database* is deprecated

This is the mapping to the terraform resources
* `Advanced data security` on SQL *server* - `azurerm_mssql_server_security_alert_policy`
* `Vulnerability Assessment Settings` on SQL *server* - `azurerm_mssql_server_vulnerability_assessment`
* Database basline settings - `azurerm_mssql_database_vulnerability_assessment_rule_baseline`

benjamin37 commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,276 @@
+  depends_on = [
+    azurerm_sql_database.test,
+    azurerm_mssql_server_security_alert_policy.test
+  ]
```
So this might be a little bit confusing, I already created a support request on Azure side and they confirmed that this is a bug, but they provided no fixing date for now.

* As I explained above the `azurerm_mssql_server_security_alert_policy.test` resource needs to exist before creating `azurerm_mssql_server_vulnerability_assessment` so this is fine
* The dependency to `azurerm_sql_database.test` should actually not be required. You can create the `azurerm_mssql_server_vulnerability_assessment` resource without having the database. The problem was that you needed to run 1 vulnerability scan on the database or do the `azurerm_mssql_server_vulnerability_assessment` setting after creating the database before you are able to set a baseline setting. I just tried it out again and this seems to work now. So I could remove the database dependency here. So the bug seems to be fixed. Shall I remove it?

benjamin37 replied:
I have removed the database dependency (as well in the documentation), hope this is fine for you.

benjamin37 commented on azurerm/resource_arm_mssql_server_vulnerability_assessment_test.go:
```diff
@@ -0,0 +1,163 @@
+
+  depends_on = [
+    azurerm_mssql_server_security_alert_policy.test
+  ]
```
see comment above, vulnerability settings on the server are a part of the security alert policy which needs to be enabled to do these settings.

benjamin37 commented on azurerm/resource_arm_mssql_server_vulnerability_assessment.go:
```diff
@@ -0,0 +1,252 @@
+		props.RecurringScans = expandRecurringScans(d)
+	}
+
+	return &vulnerabilityAssessment
```
done

benjamin37 commented on azurerm/resource_arm_mssql_server_vulnerability_assessment.go:
```diff
@@ -0,0 +1,252 @@
+		props.RecurringScans = expandRecurringScans(d)
+	}
+
+	return &vulnerabilityAssessment
```
done

benjamin37 commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline.go:
```diff
@@ -0,0 +1,223 @@
+		return err
+	}
+
+	// Currently REST API always returns "Default" as baseline that's why the Id is created manually here as a workaround
```
This bug seems to be fixed meanwhile, I'll take this manual id creation out.

benjamin37 replied:
done

benjamin37 commented:
Hi @katbyte 

I replied to your comments. I've got a question. The travis job `make docscheck` fails with the following error. I'm not sure what's missing though, can you tell me?

> Doc is missing a subcategory: website/docs/r/mssql_database_vulnerability_assessment_rule_baseline.html.markdown
> Doc is missing a subcategory: website/docs/r/mssql_server_security_alert_policy.html.markdown
> Doc is missing a subcategory: website/docs/r/mssql_server_vulnerability_assessment.html.markdown

katbyte commented:
Sorry about that the output from that tool is not the best. I've [updated it](https://github.com/terraform-providers/terraform-provider-azurerm/pull/4845) with a new PR but what you need to do is add a `subcategory` line in the yaml block at the top of the file. If you pull from master you will see it on all the other PRs or you could check #4792

benjamin37 commented:
Hi @katbyte,

thanks for the subcategory hint, I added the resources now to the Database subcategory.

katbyte commented on azurerm/resource_arm_mssql_server_security_alert_policy.go:
```diff
@@ -0,0 +1,289 @@
+						"Access_Anomaly",
+						"Data_Exfiltration",
+						"Unsafe_Action",
+					}, true),
```
Should this be case insensitive? we typically don't ignore case

benjamin37 replied:
There was no particular reason to have the case insensitive here, just for user friendly input. Removed it again.

katbyte commented on azurerm/resource_arm_mssql_server_security_alert_policy.go:
```diff
@@ -0,0 +1,289 @@
+			"state": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: suppress.CaseDifference,
```
How come we are ignoring case here? is the API returning the wrong casing compared to the constants in the SDK?

benjamin37 replied:
When I remember correctly this was the issue back then. Now it doesn't seem to be a problem anymore, removed it.

katbyte commented on azurerm/resource_arm_mssql_server_vulnerability_assessment.go:
```diff
@@ -0,0 +1,246 @@
+	serverName := id.Path["servers"]
+
+	_, err = client.Delete(ctx, resourceGroupName, serverName)
+	if err != nil {
```
These two lines could be merged:`
```
	if _, err = client.Delete(ctx, resourceGroupName, serverName); err != nil {
```

benjamin37 replied:
done

katbyte commented on website/docs/r/mssql_database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,131 @@
+
+* `baseline_result` - (Required) List of nested baseline result definitions. Each baseline result contains the following properties:
+
+  * `result` - (Required) A list representing a result of the baseline.
```
```suggestion
* `result` - (Required) A list representing a result of the baseline.
```

benjamin37 replied:
I put the documentation in a separate block documentation. Is that fine for you?

katbyte replied:
👍

benjamin37 commented:
@katbyte,

sounds like a good idea to enforce the dependency with an id of the needed resource. I'll change the resources accordingly and let you know when I'm finished.

benjamin37 commented:
Hi @katbyte,

I've made the changes, you can have an other look on the PR.

Kind regards
Benjamin

katbyte commented on azurerm/resource_arm_mssql_server_vulnerability_assessment_test.go:
```diff
@@ -0,0 +1,158 @@
+func testAccAzureRMMssqlServerVulnerabilityAssessment_server(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
could we make this
```suggestion
  name     = "acctestRG-mssql-%d"
```

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_server_security_alert_policy_test.go:
```diff
@@ -0,0 +1,157 @@
+func testAccAzureRMMssqlServerSecurityAlertPolicy_server(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
could we make this
```suggestion
  name     = "acctestRG-mssql-%d"
```

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_server_security_alert_policy_test.go:
```diff
@@ -0,0 +1,157 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMMssqlServerSecurityAlertPolicy_basic(t *testing.T) {
```
Same here could we have both a basic and update test?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_server_security_alert_policy.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	_, err = client.Get(ctx, resourceGroupName, serverName)
+	if err != nil {
```
could we merge these two lines?

benjamin37 replied:
done

katbyte commented on azurerm/resource_arm_mssql_database_vulnerability_assessment_rule_baseline_test.go:
```diff
@@ -0,0 +1,260 @@
+func testAccAzureRMMssqlDatabaseVulnerabilityAssessmentRuleBaseline_database(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
could we make this
```suggestion
  name     = "acctestRG-mssql-%d"
```

benjamin37 replied:
done

benjamin37 commented on website/docs/r/mssql_database_vulnerability_assessment_rule_baseline.html.markdown:
```diff
@@ -0,0 +1,123 @@
+---
```
done

benjamin37 commented on azurerm/resource_arm_mssql_server_vulnerability_assessment_test.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMMssqlServerVulnerabilityAssessment_basic(t *testing.T) {
```
done

benjamin37 commented:
Hi @katbyte,

did some changes you requsted. Still not fixed the `panic: Invalid address to set` error, why shouldn't I call the set method to set the resource group in the read? This seems to be necessary to me. I will have an other look on this bug tomorrow!

About giving you the rights to push on the branch I will also check back tomorrow.

Kind regards
Benjamin

katbyte commented:
@benjamin37, there is a bug on master right now causing acctests to fail.

As for the `d.Set()`, your right in that it should be required there 🤔 maybe i read the error wrong. Once we fix master we can see if its still there.

benjamin37 commented:
Hi @katbyte,

thanks for the hint, this saved me some time :-)
I rebased against master again and run all the tests. They seem to work fine now.

Sadly I don't have the permissions to let you push on the branch of this PR.

If there is anything to do on this PR please let me know!

Kind regards
Benjamin

benjamin37 commented:
Hi @katbyte,

ups, that's obviously a bug, I changed it to use the correct read function.

Kind regards
Benjamin
